5.5/mysql/bin/run-mysqld.sh

```diff
@@ -1,43 +1,96 @@
-	echo "  \$MYSQL_PASSWORD"
-	echo "  \$MYSQL_DATABASE"
+	echo "  \$MYSQL_USER (regex: '$mysql_identifier_regex')"
+	echo "  \$MYSQL_PASSWORD (regex: '$mysql_password_regex')"
+	echo "  \$MYSQL_DATABASE (regex: '$mysql_identifier_regex')"
+	echo "Optional:"
+	echo "  \$MYSQL_ROOT_PASSWORD (regex: '$mysql_password_regex')"
 	exit 1
 }
 
-test -z "$MYSQL_USER" && usage
-test -z "$MYSQL_PASSWORD" && usage
-test -z "$MYSQL_DATABASE" && usage
-
-if [ ! -d '/var/lib/mysql/data' ]; then
-
-	echo 'Running mysql_install_db ...'
-	scl enable mysql55 "mysql_install_db --user=mysql --datadir=/var/lib/mysql/data"
-	echo 'Finished mysql_install_db'
-
-	# These statements _must_ be on individual lines, and _must_ end with
-	# semicolons (no line breaks or comments are permitted).
-	# TODO proper SQL escaping on ALL the things D:
-	TEMP_FILE='/tmp/mysql-first-time.sql'
-	cat > "$TEMP_FILE" <<-EOSQL
-		DELETE FROM mysql.user ;
-		CREATE USER '${MYSQL_USER}'@'%' IDENTIFIED BY '${MYSQL_PASSWORD}' ;
-		GRANT ALL ON *.* TO '${MYSQL_USER}'@'%' WITH GRANT OPTION ;
-		DROP DATABASE IF EXISTS test ;
-		CREATE DATABASE IF NOT EXISTS \`${MYSQL_DATABASE}\` ;
-		FLUSH PRIVILEGES ;
-	EOSQL
+function valid_mysql_identifier {
+	local var="$1" ; shift
+	[[ "${var}" =~ $mysql_identifier_regex ]]
+}
 
-	set -- "$@" --init-file="$TEMP_FILE"
-fi
+function valid_mysql_password {
+	local var="$1" ; shift
+	[[ "${var}" =~ $mysql_password_regex ]]
+}
+
+valid_mysql_identifier "$MYSQL_USER"     || usage
+valid_mysql_password   "$MYSQL_PASSWORD" || usage
+valid_mysql_identifier "$MYSQL_DATABASE" || usage
 
-chown -R mysql:mysql /var/lib/mysql
+# Make sure env variables don't propagate to mysqld process.
+mysql_user="$MYSQL_USER" ; unset MYSQL_USER
+mysql_pass="$MYSQL_PASSWORD" ; unset MYSQL_PASSWORD
+mysql_db="$MYSQL_DATABASE" ; unset MYSQL_DATABASE
+
+# Root password.
+if [ "$MYSQL_ROOT_PASSWORD" ]; then
+	valid_mysql_password "$MYSQL_ROOT_PASSWORD" || usage
+	root_pass="$MYSQL_ROOT_PASSWORD"
+fi
+unset MYSQL_ROOT_PASSWORD
 
 # SCL in CentOS/RHEL 7 doesn't support --exec, we need to do it ourselves
-export X_SCLS="mysql55"
-source /opt/rh/mysql55/enable
+# The '|| exit 1' is here so -e doesn't propagate into scl_source.
+source scl_source enable mysql55 || exit 1
+
+# Poll until MySQL responds to our ping.
+function wait_for_mysql {
+	pid=$1 ; shift
+
+	while [ true ]; do
+		if [ -d "/proc/$pid" ]; then
+			mysqladmin --socket=/tmp/mysql.sock ping &>/dev/null && return 0
+		else
+			return 1
+		fi
+		echo "Waiting for MySQL to start"
+		sleep 1
+	done
+}
+
+if [ ! -d '/var/lib/mysql/mysql' ]; then
+
+	echo 'Running mysql_install_db'
+	mysql_install_db --user=mysql --datadir=/var/lib/mysql
+	# TODO: Not needed with --user=mysql. However, we should
+	#       strive to make this script runnable without root privileges
+	#chown -R mysql:mysql /var/lib/mysql
+
+	# Now start mysqld and add appropriate users.
+	echo 'Starting mysqld to create users'
+	/opt/rh/mysql55/root/usr/libexec/mysqld \
```

Comment:
Just because of one or two execs? No.. It's libexec, it shouldn't be in PATH, IMO.